crates/sui-core/src/consensus_adapter.rs

```diff
@@ -225,92 +212,87 @@ impl ConsensusAdapter {
         // refreshed frequently enough to make sure this is Byzantine-resistant
     }
 
-    /// Submit a transaction to consensus, wait for its processing, and notify the caller.
-    // Use .inspect when its stable.
-    #[allow(clippy::option_map_unit_fn)]
-    pub async fn submit(&self, certificate: &VerifiedCertificate) -> SuiResult {
-        let processed_waiter = self
-            .authority
-            .consensus_message_processed_notify(certificate.digest());
-        // Serialize the certificate in a way that is understandable to consensus (i.e., using
-        // bincode) and it certificate to consensus.
-        let transaction = ConsensusTransaction::new_certificate_message(
-            &self.authority.name,
-            certificate.clone().into(),
-        );
-        let tracking_id = transaction.get_tracking_id();
-        let tx_digest = certificate.digest();
-        debug!(
-            ?tracking_id,
-            ?tx_digest,
-            "Certified transaction consensus message created"
-        );
+    /// This method blocks until transaction is persisted in local database
+    /// It then returns handle to async task, user can join this handle to await while transaction is processed by consensus
+    ///
+    /// This method guarantees that once submit(but not returned async handle) returns,
+    /// transaction is persisted and will eventually be sent to consensus even after restart
+    pub async fn submit(
+        self: &Arc<Self>,
+        transaction: ConsensusTransaction,
+    ) -> SuiResult<JoinHandle<()>> {
+        let _lock = if transaction.is_user_certificate() {
+            let lock = self.authority.get_reconfig_state_read_lock_guard().await;
+            if !lock.should_accept_consensus_certs() {
```

Comment:
I think you meant should_accept_user_certs here? Sorry if my namings are not clear.

Reply:
Indeed, good catch!